make ECDH unpadding more robust

hko-s opened the pull request:
The padding mandated by [RFC 6637](https://www.rfc-editor.org/info/rfc6637) may exceed the blocksize of 8 bytes. This caused problems, for example with messages from GOpenPGP, which use AES128 secret keys and contain 21 bytes of padding: the previous implementation failed to decrypt these and returned an Error::UnpadError.

hko-s commented:
My impression is that, while RFC 6637 references PKCS5 as "the general idea" that the padding in OpenPGP follows, using a PKCS5 unpadding function will fail, because the length of the padding used in OpenPGP can exceed the block size of 8.
So it's really *not* PKCS5 padding, in a stricter sense, if I understand correctly.

However, it seems innocent enough to implement the RFC 6637 unpadding scheme manually, as outlined in this PR.

hko-s commented:
We could consider, in a separate PR, to use ECDH padding to conceal key size (as outlined in RFC 6637, for example with AES. this means always padding to 40 bytes). Currently the padding in rpgp only pads to block size. Which is a valid approach, and can save a few bytes of space, but gives eavesdroppers more insight into the key length.

I have no strong feelings about this question, right now, but wanted to raise it while the padding is on my mind. If there is interest, I can open a separate ticket to track this issue.

hko-s commented on src/crypto/ecdh.rs:
```diff
@@ -145,18 +145,41 @@ pub fn decrypt(priv_key: &ECDHSecretKey, mpis: &[Mpi], fingerprint: &[u8]) -> Re
+        let pad = decrypted_key_padded.last().expect("is not empty");
+
+        // Padding length seems to exceed size of the padded message
+        if *pad as usize > len {
```
I believe the padding scheme in RFC 6637 explicitly allows the padding to exceed block length. The rationale is to obfuscate the length of the symmetric key.

For example, RFC 6637 mentions 21 bytes of padding as a legal format for an AES128 key (also see my comment in https://github.com/rpgp/rpgp/pull/280/files#diff-962d2eaeb82df8d56721a98d7d86e270a5021cfa2b361b25146ecbd68d47cf51R150).

The unpadded key must have the expected length for the algorithm, of course. The previous code didn't check for this property. I'm not sure if it would be useful to check for that.

My feeling is: this unpadding function should not fail with panics, but if the sender sends us a padded key that is just 1 byte long, then that's what we have. Trying to us that key should fail later in the code, then?

hko-s replied:
Putting a finer point on it: The purpose of this PR is to handle padding that exceeds block length.

Because that currently fails, which causes interop problems when using AES128 with GopenPGP, as well as with bouncy castle-based implementations (like PGPainless).

dignifiedquire replied:
> The unpadded key must have the expected length for the algorithm, of course. The previous code didn't check for this property. I'm not sure if it would be useful to check for that. 

we really should

dignifiedquire replied:
re reading the RFC

>  This encoding allows
   the sender to obfuscate the size of the symmetric encryption key used
   to encrypt the data.  For example, assuming that an AES algorithm is
   used for the session key, the sender MAY use 21, 13, and 5 bytes of
   padding for AES-128, AES-192, and AES-256, respectively, to provide
   the same number of octets, 40 total, as an input to the key wrapping
   method. 

it seems that the padding is not variable but deterministic, depending on the key size, so that you always end up with 40 bytes

hko-s replied:
> it seems that the padding is not variable but deterministic, depending on the key size, so that you always end up with 40 bytes (if the obfuscation is used)

I don't think that's what happens in practice. As far as I recall, not all implementation produce padded AES keys at 40 bytes length.

Note that it's just a "MAY", not a "MUST". Implementations do different padding, but they are allowed to pad *up to* the maximum length.

dignifiedquire replied:
sigh, I hate variable padding…

r10s commented:
just want to say hello 👋 - i am following these recent rpgp ongoings and high-level discussions with admiration. super-nice to see such a push here!

dignifiedquire commented:
can you add a test with one of those padded keys please, and also file an issue to do the obfuscation impl (to 40 bytes..gotta keep some sanity)

wiktor-k commented on src/crypto/ecdh.rs:
```diff
@@ -145,18 +145,41 @@ pub fn decrypt(priv_key: &ECDHSecretKey, mpis: &[Mpi], fingerprint: &[u8]) -> Re
-        let unpadded_last_block = Pkcs7::unpad(last_block)?;
-        let unpadded_len = offset + unpadded_last_block.len();
+        // The last byte should contain the padding symbol, which is also the padding length
+        let pad = decrypted_key_padded.last().expect("is not empty");
```
I'm just wondering if it's OK to panic in a library even if the data is malformed? :thinking: (I was thinking about using `return Err(Error::UnpadError);`). Just a random note. :wave:

hko-s replied:
This code should probably be rewritten to be more readable. But that said, as far as I can tell, this `expect()` cannot trigger:

The `ensure!(!decrypted_key_padded.is_empty(), "empty key is not valid");` right above makes sure that there is at least one byte of data in `decrypted_key_padded`.

wiktor-k replied:
Ok. That clarifies it for me, although it looked weird at first... :+1: (closing this thread)

hko-s commented:
The decrypted (and padded) key in this message reads:

```
[07,   # AES with 128-bit key

# AES 128 session key:
  e1, fc, 84, f9, 2f, 9e, 85, 01,
  11, fb, 12, ee, ba, 1b, e3, e2,

# checksum:
  09, 53,

# padding:
  15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15]
```

There are 21 bytes of padding in the message, padding it to a total length of 40 bytes.

hko-s commented:
By contrast, when encrypting to Alice's key with `sqop`, we get a message that uses AES256 and a padded session key with only 5 bytes of padding:

```
-----BEGIN PGP MESSAGE-----

wV4D4uDroNMMT4QSAQdAxYnDMbJRKbO9/XbXVr7hxq1TbIZjkMD4no/tF/uLJEsw
/7/4n/oGWofvF7YbqEc7fwIdAt5KY7GRWTLvT4/KOt9280OjITmDvLm4lxsiPj7e
0jcBBFNzsp79iSX+rmLS4sT/Fd1JOPnqBY5FhxPfcZy/xFNuDC7LMYTZVQJRbkCn
UheUc5ALR3Fw
=9v3V
-----END PGP MESSAGE-----
```

Decrypted and padded key:

```
 [09,

 01, 36, 4e, 82, c4, be, 22, fb,
 2c, 20, 9f, 25, 6a, 73, cd, e4,
 9b, 2a, 27, d6, f0, 1e, 59, 1f,
 e1, 29, fd, 9e, 02, 28, 94, 79,

 0e, 68,

 05, 05, 05, 05, 05]
```

wiktor-k commented:
> In preparation to making progress on this, I'm generating and documenting artifacts.

Don't "document" it, just add test cases with these vectors! This will be a great regression test suite that will be always checked by the CI! 😌

hko-s commented:
> > In preparation to making progress on this, I'm generating and documenting artifacts.
> 
> Don't "document" it, just add test cases with these vectors! This will be a great regression test suite that will be always checked by the CI! 😌

Hah :) yes, of course! This is my scratchpad, so I can do little bits of work, get interrupted, and resume later (writing up things publicly seems useful, if I have to write them up anyway)

hko-s commented:
> can you add a test with one of those padded keys please, and also file an issue to do the obfuscation impl (to 40 bytes..gotta keep some sanity)

- I opened #292 to track extending the padding that rpgp emits,
- added two tests that decrypt ECDH messages that use 5 and 21 bytes of padding respectively (I confirmed that the second test fails without this PR), and
- added sanity checking after unpadding, to `ensure!` that the format of the unpadded key meets our expectations.

hko-s commented:
> This currently fails `crypto::ecdh::tests::test_encrypt_decrypt`, not ready for review yet?

The PR definitely requires input (but I admit I was too optimistic, and thought there would be no more remaining problems/questions :)).

What CI complains about is this test: https://github.com/hko-s/rpgp/blob/ecdh-unpad/src/crypto/ecdh.rs#L318-L324
The test encrypts and decrypts plaintexts of varying length and with arbitrary (random) content.

[Here](https://github.com/rpgp/rpgp/pull/280#discussion_r1482809283), @dignifiedquire suggested we do more checks on the format of the unpadded data. I [implemented checks](https://github.com/hko-s/rpgp/blob/ecdh-unpad/src/crypto/ecdh.rs#L188-L205) that make assumptions about format of the decrypted/unpadded data.
However, `crypto::ecdh::tests::test_encrypt_decrypt` unit test seems to assume that `pgp::crypto::ecdh::decrypt` can handle arbitrary plaintexts. If that's what we want, then I don't think we can do stricter checks on the unpadded plaintext.

@dignifiedquire can you have a look at my check? I'm not sure if I implemented roughly what you had in mind, or if your idea was entirely different.

dignifiedquire commented:
was thinking more of length checks, tbh I think it is enough to just keep the "not empty" check, and leave things as they are.
